distributed/deploy/tests/test_runner.py

```diff
@@ -0,0 +1,31 @@
+import asyncio
+from contextlib import suppress
+from time import time
+
+import pytest
+
+from dask.distributed import Client
+
+from distributed.deploy.runner import AsyncCommWorld, AsyncRunner
+from distributed.utils_test import loop  # noqa: F401
+
+
+@pytest.mark.asyncio
+@pytest.mark.timeout(10)
+async def test_runner(loop):
+    commworld = AsyncCommWorld()
+
+    async def run_code(commworld):
+        with suppress(SystemExit):
+            async with AsyncRunner(commworld, asynchronous=True) as runner:
+                async with Client(runner, asynchronous=True) as c:
+
+                    start = time()
+                    while len(c.scheduler_info()["workers"]) != 2:
+                        assert time() < start + 10
+                        await asyncio.sleep(0.2)
+
+                    assert await c.submit(lambda x: x + 1, 10).result() == 11
+                    assert await c.submit(lambda x: x + 1, 20).result() == 21
+
+    await asyncio.gather(*[run_code(commworld) for _ in range(4)])
```

Comment:
IIUC, this spawns four async tasks (or equivalently four MPI jobs, four processes, etc.) of which one decides to become a scheduler, two decide to become a worker and one will become a client.
Scheduler and worker will effectively ignore the ctx mgr body and act as if they were ordinary server nodes. While the client will actually connect and execute the code. So, effectively this is an async local cluster with two workers in disguise.

This implementation will negotiate the "role" (which I was calling node type) via `AsyncCommWorld` which in this case is simply an async lock but in general this interface would need to implement some sort of leader election via a shared filesystem, a distribtued KV store (zookeeper, etcd,...). This inteface is currently discussed in https://github.com/dask/distributed/issues/4715

Is this an accurate summary of what's going on?

Reply:
Yep that's it!

Reply:
> The key difference between this and a Cluster is that the roles are worked out after the processes 

From an implementation perspective, the two may look alike but conceptually I feel these are very different concepts which is why I had a hard time following.

Conceptually I see the `Cluster` as the entity which is allowed, or even required, to talk to the _hardware_ or _resource manager_ (trying to avoid the "cluster" term here). It will talk to this hardware manager via some kind of API to spawn new `ServerNode` instances of type `{Scheduler|Worker}` _somewhere else_ (different process, different VM, ...) while the `Runner` will spawn *one* `ServerNode` _next to it_, similar to what the `Nanny` does with the `Worker`.

Something like
![image](https://user-images.githubusercontent.com/8629629/115591765-c19c2600-a2d2-11eb-9cdf-afc39d000b17.png)

Anyhow, I think that's a nice concept, it just took me while to understand and I think it is valuable to document this properly.
I'm wondering if we need some standardized interface for `AsyncCommWorld` or whether this is too backend specific

Reply:
> Anyhow, I think that's a nice concept, it just took me while to understand and I think it is valuable to document this properly.

Great. I think this is useful, but may not be the final incarnation of it. I had a chat with @mrocklin yesterday about service discovery and leadership election which is related to this change. 

> I'm wondering if we need some standardized interface for AsyncCommWorld or whether this is too backend specific

I think that may be too specific. That class was more of a necessary evil because the lock had to live somewhere.